Source/Particles/Collision/BinaryCollision/SingleNuclearFusionEvent.H

```diff
@@ -0,0 +1,185 @@
+                               const amrex::Real& probability_threshold,
+                               const amrex::Real& probability_target_value,
+                               const NuclearFusionType& fusion_type,
+                               const amrex::RandomEngine& engine)
+{
+    // General notations in this function:
+    //     x_sq denotes the square of x
+    //     x_star denotes the value of x in the center of mass frame
+
+    const amrex::ParticleReal w_min = amrex::min(w1, w2);
+    const amrex::ParticleReal w_max = amrex::max(w1, w2);
+
+    constexpr auto one_pr = amrex::ParticleReal(1.);
+    constexpr auto inv_two_pr = amrex::ParticleReal(1./2.);
+    constexpr auto inv_four_pr = amrex::ParticleReal(1./4.);
+    constexpr amrex::ParticleReal c_sq = PhysConst::c * PhysConst::c;
+    constexpr amrex::ParticleReal inv_csq = one_pr / ( c_sq );
+
+    const amrex::ParticleReal m1_sq = m1*m1;
+    const amrex::ParticleReal m2_sq = m2*m2;
+
+    // Compute Lorentz factor gamma in the lab frame
+    const amrex::ParticleReal g1 = std::sqrt( one_pr + (u1x*u1x+u1y*u1y+u1z*u1z)*inv_csq );
+    const amrex::ParticleReal g2 = std::sqrt( one_pr + (u2x*u2x+u2y*u2y+u2z*u2z)*inv_csq );
+
+    // Compute momenta
+    const amrex::ParticleReal p1x = u1x * m1;
+    const amrex::ParticleReal p1y = u1y * m1;
+    const amrex::ParticleReal p1z = u1z * m1;
+    const amrex::ParticleReal p2x = u2x * m2;
+    const amrex::ParticleReal p2y = u2y * m2;
+    const amrex::ParticleReal p2z = u2z * m2;
+    // Square norm of the total (sum between the two particles) momenta in the lab frame
+    const amrex::ParticleReal p_total_sq = std::pow(p1x+p2x, 2) +
+                                     std::pow(p1y+p2y, 2) +
+                                     std::pow(p1z+p2z, 2);
+
+    // Total energy in the lab frame
+    const amrex::ParticleReal E_lab = (m1 * g1 + m2 * g2) * c_sq;
+    // Total energy squared in the center of mass frame, calculated using the Lorentz invariance
+    // of the four-momentum norm
+    const amrex::ParticleReal E_star_sq = E_lab*E_lab - c_sq*p_total_sq;
+
+    // Kinetic energy in the center of mass frame
+    const amrex::ParticleReal E_kin_star = std::sqrt(E_star_sq) - (m1 + m2)*c_sq;
+
+    // Compute fusion cross section as a function of kinetic energy in the center of mass frame
+    auto fusion_cross_section = amrex::ParticleReal(0.);
+    if (fusion_type == NuclearFusionType::ProtonBoron)
+    {
+        fusion_cross_section = ProtonBoronFusionCrossSection(E_kin_star);
+    }
+
+    // Square of the norm of the momentum of one of the particles in the center of mass frame
+    // Formula obtained by inverting E^2 = p^2*c^2 + m^2*c^4 in the COM frame for each particle
+    const amrex::ParticleReal p_star_sq =
+            E_star_sq*inv_four_pr*inv_csq - (m1_sq + m2_sq)*c_sq*inv_two_pr +
+            std::pow(c_sq, 3)*inv_four_pr*std::pow(m1_sq - m2_sq, 2)/E_star_sq;
```

Comment:
I agree that it's not obvious. My plan was to add at some point a detailed derivation in the developers documentation so that we have this in writing somewhere.

But basically the idea is that the energy of particle 1 (2) in the COM frame is E1 = sqrt(p²c² + m1²c⁴) (E2 = sqrt(p²c² + m2²c⁴)). I can now express E1 + E2 = `E_star_sq` as a function of p² = `p_star_sq`. And by inverting this formula, I can get `p_star_sq` as a function of `E_star_sq`, which is what is done here.

I'm not sure what else I can add as a comment here without going into too much details. My plan was to add something like `(see developers documentation for more details)` once we actually have something in the documentation. Hopefully, that will be enough.